openc3/spec/local_s3.rb

```diff
@@ -0,0 +1,226 @@
+
+#      h_obj = Aws::S3::Types::HeadObjectOutput.new()
+      true
+    rescue Exception => e
+       puts "HEAD_OBJECT: #{e.message}"
+       raise Aws::S3::Errors::NotFound.new(nil, "not found")
+    end
+
+    def list_objects(args)
+      resp = list_objects_v2(args)
+      resp.contents
+    end
+
+    def list_objects_v2(args)
+      bucket = get_bucket(args[:bucket])
+      raise Aws::S3::Errors::NoSuchBucket.new(nil, nil) if bucket.nil?
+      count = 0
+      max = args[:max_keys].to_i
+      max = 1000 unless (max.between?(1, 1000))
+      is_truncated = false
+      con_token = next_token = args[:continuation_token] # ???
+      cmn_pfxs = {}
+      # bucket query
+      fullpath = File.expand_path(File.join(@fs_root, bucket.name))
+      filenames = Dir.glob(File.join(fullpath, "**/*")).sort
+      if (con_token)
+        filenames = filenames.sort.select {|fname| !(fname.nil?) && fname >= con_token}
+      end
+      contents = []
+      filenames.each do |fname|
+        if (!File.directory?(fname))
+          pattern = "\\S*#{args[:bucket]}\/(\\S*)" # get only bucket thru filename
+          obj_key = ""
+          if (match = fname.match(pattern))
+            obj_key = match.captures[0]
+          end
+          b_name = File.basename(fname)
+          pattern = "/\S*#{args[:bucket]}(\S*)#{b_name}/" # get only between bucket and basename
+          case b_name
+          when (count > max)
+            next_token = fname
+            is_truncated = true
+            break
+          when '.', '..'
+          # just fall through
+          else
+            count += 1
+            s3_obj = S3Object.new(bucket_name: args[:bucket], key: obj_key)
+            contents << s3_obj
+            prefix = File.dirname(fname).match(pattern).to_s
+            commoner = ''
+            prefix.split(args[:delimiter]).each do |chunk|
+              commoner += '/'+chunk
+              cmn_pfxs[commoner] = 1
+            end
+          end
+        end
+      end
+      response = Struct.new(:contents, :is_truncated, :next_continuation_token, :common_prefixes)
+      resp = response.new(contents, is_truncated, next_token, cmn_pfxs.keys.sort)
+      return resp
+    end
+
+    def put_bucket_policy(*args)
+      # no-op: this space intentionally left blank
+    end
+
+    # MEMO: raise ::Aws::S3::Errors::NoSuchKey.new(nil, nil) if s3_obj.nil?
+
+    def put_object(bucket:, key:, body:, content_type:, cache_control:, metadata:, checksum_algorithm:)
+      bucket_obj = get_bucket(bucket)
+      if !bucket_obj
+        # Lazily create a bucket.  Make this return the proper error.
+        bucket_obj = create_bucket(bucket)
+      end
+      # ignores args[:metadata] for now
+
+      fullpath = File.expand_path(File.join(@fs_root, bucket, key))
+      FileUtils.mkdir_p(File.dirname(fullpath))
+      File.open(fullpath, 'wb') do |file|
+        file.write(body)
+      end
+    end
+
+    def get_objects(args)
```

Comment:
Is this not needed?

Reply:
That is correct: it is not actually needed.  Removed.